### ISSUE TITLE
[TySan] False positive related to structs in structs?

### ISSUE THREAD

**Issue by seanm** [llvm/llvm-project#122517]
Maybe I'm just not using it right, or not understanding something, but OTOH TySan is new, and thus probably a bit buggy... I used creduce to create a C test case showing, I think, a false positive.

```C
typedef struct {
	int len;
	char *list;
} str_list;

typedef struct {
	str_list infiles;
	char command;
} nt_opts;

int add_string(str_list *slist) {
	(void)(slist->list);
}

int process_opts(nt_opts *opts) {
	add_string(&opts->infiles);
}

void free_opts_mem(nt_opts *nopt) {
	(void)(nopt->infiles.list); // •••TySan warns here•••
}

void main(int argc, char *argv[]) {
	nt_opts opts;
	int rv = process_opts(&opts);
	free_opts_mem(&opts);
}
```

then I run:

`(xcrun /Users/sean/llvm/llvm-install/bin/clang -w -g -fsanitize=type test.c && ./a.out)`

and I get:

```
==4848==ERROR: TypeSanitizer: type-aliasing-violation on address 0x00016cfc71f0 (pc 0x000102e3b690 bp 0x00016cfc6ea0 sp 0x00016cfc6620 tid 24776816)
READ of size 8 at 0x00016cfc71f0 with type p1 omnipotent char (in <anonymous type> at offset 8) accesses an existing object of type p1 omnipotent char (in <anonymous type> at offset 8)
    #0 0x000102e3b68c in free_opts_mem test.c:20
```

Aside:

- what is an "omnipotent char"?
- what is "type p1" referring to?

**Comment by seanm** [comment]
@fhahn @gbMattN this may interest you.

Perhaps it is related to https://github.com/llvm/llvm-project/pull/120412 afterall?

**Comment by seanm** [comment]
I forget to say that the clang I built is from the https://github.com/llvm/llvm-project/pull/120412 PR.

**Comment by seanm** [comment]
> This is the terminology Clang uses to encode the type information for LLVM, which tysan uses verbatim. We could probably try better to translate the types to something closer related to the types in source

Thanks. I've created https://github.com/llvm/llvm-project/issues/122522 to suggest some documentation ideas.

I think I'll have to stop playing with it for now, as best as I can tell at least 3/4 of the reports I'm seeing from TySan are variants of this reduced case. It's just too much noise for now.

Excited to try more later!

**Comment by gbMattN** [comment]
> Perhaps it is related to [#120412](https://github.com/llvm/llvm-project/pull/120412) afterall?

I can reproduce this on main, and it seems its fixed in [#120412](https://github.com/llvm/llvm-project/pull/120412) which will be merged in as soon as the build tests pass!

**Comment by seanm** [comment]
Confirmed fixed. That got me down from 827 `ERROR: TypeSanitizer` to just 61 now.

The first of the 61 looks like a false positive to me too... I'll creduce it now...

**Comment by seanm** [comment]
> The first of the 61 looks like a false positive to me too... I'll creduce it now...

https://github.com/llvm/llvm-project/issues/122934
